src/memkind_mem_attributes.c

```diff
@@ -291,18 +302,67 @@ int get_mem_attributes_hbw_nodes_mask(struct bitmask **hbw_node_mask)
             }
 
             if (bandwidth >= hbw_threshold) {
-                numa_bitmask_setbit(*hbw_node_mask, target->os_index);
+                int dist = numa_distance(init_node->os_index, target->os_index);
+                if (dist < min_distance) {
+                    min_distance = dist;
+                    VEC_CLEAR(&current_dest_nodes);
+                    VEC_PUSH_BACK(&current_dest_nodes, target->os_index);
+                } else if (dist == min_distance) {
+                    VEC_PUSH_BACK(&current_dest_nodes, target->os_index);
+                }
+            }
+        }
+
+        vec_size = VEC_SIZE(&current_dest_nodes);
+
+        if (vec_size == 0) {
+            status = MEMKIND_ERROR_MEMTYPE_NOT_AVAILABLE;
+            for (i = 0; i < num_cpu; ++i) {
+                if (VEC_CAPACITY(&node_arr[i]))
+                    VEC_DELETE(&node_arr[i]);
+            }
+            log_err("No HBW Nodes for init node %d.", init_node->os_index);
+            goto free_node_arr;
+        }
+
+        //validate single NUMA Node condition
+        if (node_variant == NODE_VARIANT_SINGLE &&
+            vec_size > 1) {
+            log_err("Invalid Numa Configuration for Node %d", init_node->os_index);
+            status = MEMKIND_ERROR_RUNTIME;
+            for (i = 0; i < num_cpu; ++i) {
```

Comment:
this is the same code as 320-323 - could you move it to "free_node_array" section?